packages/gatsby/src/state-machines/actions.ts

```diff
@@ -0,0 +1,113 @@
+import {
+  assign,
+  AnyEventObject,
+  ActionFunction,
+  spawn,
+  ActionFunctionMap,
+  DoneEventObject,
+} from "xstate"
+import { Store } from "redux"
+import { IBuildContext, IMutationAction } from "../services"
+import { actions } from "../redux/actions"
+import { listenForMutations } from "../services/listen-for-mutations"
+import { DataLayerResult } from "./data-layer"
+import { assertStore } from "../utils/assert-store"
+
+export const callRealApi = async (
+  event: IMutationAction,
+  store?: Store
+): Promise<unknown> => {
+  assertStore(store)
+  const { type, payload } = event
+  if (type in actions) {
```

Comment:
Is it expected that this check can fail?

Reply:
At 100k pages it's < 1s. There was an implementation of this in the inc build cli that was O(n^2), and which was the reason the 100k site initially took 4 hours to build! This one is much quicker.

Reply:
It shouldn't, but it could in theory have a redux action dispatched that doesn;t have a handler. This avoids runtime errors in that case.

Reply:
Silently so. I wonder if we could raise some kind of notification / analytics so we're aware this happens as it's very unlikely to be intentional...